### ISSUE TITLE
isEqualToComparingFieldByFieldRecursively is not symmetrical when reflection is involved

### ISSUE THREAD

**Issue by crizzis** [assertj/assertj-core#1369]
#### Summary
`isEqualToComparingFieldByFieldRecursively` is not symmetrical when a field of subclass type is populated with an object with a superclass type using reflection. 

(this may sound as an obscure use case, but such 'mistakes' tend to happen with mapping/serialization libraries, which use reflection heavily. It just happened to me with Orika, for one)

#### Example

```java
import static org.assertj.core.api.Assertions.assertThat;

import java.lang.reflect.Field;
import org.junit.Test;

public class AssertJEqualityTest {

  @Test
  public void isEqualToComparingFieldByFieldRecursively_shouldBeSymmetrical()
      throws NoSuchFieldException, IllegalAccessException {
    //given
    final ReferencingClass actual = new ReferencingClass();
    actual.setReferenced(new ChildClass());
    final ReferencingClass expected = new ReferencingClass();
    Field referencedField = ReferencingBaseClass.class.getDeclaredField("referenced");
    referencedField.setAccessible(true);
    referencedField.set(expected, new BaseClass());
    //then
    assertThat(expected).isEqualToComparingFieldByFieldRecursively(actual);
    assertThat(actual).isEqualToComparingFieldByFieldRecursively(expected);
  }
}

class BaseClass {
  private String fieldA = "fieldA";
  private String fieldB = "fieldB";

  public String getFieldA() {
    return fieldA;
  }

  public void setFieldA(String fieldA) {
    this.fieldA = fieldA;
  }

  public String getFieldB() {
    return fieldB;
  }

  public void setFieldB(String fieldB) {
    this.fieldB = fieldB;
  }
}

class ChildClass extends BaseClass {
  private String fieldC = "fieldC";

  public String getFieldC() {
    return fieldC;
  }

  public void setFieldC(String fieldC) {
    this.fieldC = fieldC;
  }
}

class ReferencingBaseClass<T extends BaseClass> {
  private T referenced;

  public T getReferenced() {
    return referenced;
  }

  public void setReferenced(T referenced) {
    this.referenced = referenced;
  }
}

class ReferencingClass extends ReferencingBaseClass<ChildClass> {
}

```

In the example above, only the second assertion fails. It is a bit similar to [this issue](https://github.com/joel-costigliola/assertj-core/issues/1054). The test case above shows, though, that one side of the comparison is favoured when AssertJ checks whether the two objects have the same fields.

**Comment by joel-costigliola** [comment]
Thanks for reporting, I'll have a look in the next days/weeks.

**Comment by joel-costigliola** [comment]
@dev-priporov Sorry for the (very) late feedback, I have given it a  try with the current code and the new recursive api (#1002). The error message is the same:
```
Path to difference: <referenced>
- actual  : <org.assertj.core.api.object.ChildClass@f5958c9>
- expected: <org.assertj.core.api.object.BaseClass@77c2494c>
- reason  : org.assertj.core.api.object.ChildClass can't be compared to org.assertj.core.api.object.BaseClass as BaseClass does not declare all ChildClass fields, it lacks these:[fieldC]
```
To me this is the correct default behavior, you can't compare a `ChildClass` to its `BaseClass` as it has a different data structure, said otherwise because `BaseClass` is not a `ChildClass` they can't be compared but the opposite works as `ChildClass` is a `BaseClass`.

If you ignore the extra field, then you can make the assertion work, ex:
```java
assertThat(actual).usingRecursiveComparison()
                  .ignoringFields("referenced.fieldC")
                  .isEqualTo(expected);
```

Since I believe the default behavior is correct, I think the only way to support this use case is to allow ignoring extra actual fields or going one step further limit the comparison to common fields only. 

Having said that, it feels to me this is an edge case because it looks like you don't fully know what fields are being compared and if you know that, then why not ignoring them ?

**Comment by crizzis** [comment]
@joel-costigliola since you've commented on whether or not the issue is in fact relevant, I assume you meant to ping me instead (it I'm mistaken, feel free to ignore my comment). 

First of all, thanks and no worries for the late reply, this is indeed an edge case as you pointed out in your comment ;)

That being said, I think the entire issue boils down to two questions: 

1. What should happen when a subclass with an extra field is compared to a superclass
2. Is it ever OK that `assertThat(expected).isEqualToXXX(actual)` and `assertThat(actual).isEqualToXXX(expected)` yield different results

The issue I see is not that the second assertion fails. The issue is that the two assertions do not behave the same way, even though the only difference is the order of the inputs to the comparison. 

In other words, while I think #1 is debatable, #2 shouldn't really be. IMHO, an `isEqualToXXX` assertion suddenly changing behavior just because you swap `actual` with `expected` is a violation if the principle of least astonishment: comparisons are by intuition symmetrical. You would expect that whether to put `actual` in `assertThat(...)` and `expected` in `isEqualTo(...)` or the other way around is a stylistic choice when, in reality, it changes the behavior of the test. 

At the end of the day, if `assertThat(expected).isEqualTo(actual)` passes, and `assertThat(actual).isEqualTo(expected)` fails, what does that tell me? Are the objects equal or not? ;)

**Comment by joel-costigliola** [comment]
I agree with what you say if the the compared objects have the same type but when comparing subclass instance with a super class instance (ColorPoint and Point for example) there is no way to have a proper equals implementation (see Effective Java - Obey the general contract when overriding equals for a detailed explanation). 

The recursive comparison has the same "issue" which we addressed by not guaranteeing the symmetry.

**Comment by crizzis** [comment]
@joel-costigliola Well, my reading of Effective Java is that you shouldn't try to maintain any sort of compatibility of `equals` between superclasses and subclasses precisely because otherwise the symmetry would break. If two objects are not of the exact same class, they cannot possibly be equal. By the same token, I would have expected both of the assertions from the example to fail, not just one of them. 

But I suppose we have to agree to disagree. My only issue is that when `expected` is an object I contruct, and `actual` comes from a declarative mapping library, I cannot know what I should expect to find in `actual`. How do I make an informed decision which object to put in `assertThat(...)` and which in `.isEqualToXxx(...)`? It would seem I need both assertions, and I wish the assymetry of the comparison was better documented.
